### PR TITLE
Fast tips

### PR DESCRIPTION
**Pull request by CPilbrow**
Adds a cache to get-group-tangle to avoid recalculation.

### CODE REVIEW

File: lib/get-group-tangle.js [hunk 1]
```diff
@@ -1,18 +1,59 @@
...
+        // Add message to Reduce
+        if (tangles && tangles.group && tangles.group.previous) {
+          cache[msgGroupId].addNodes([{ key: msg.key, previous: tangles.group.previous }]) // Get key and previous from msg
+        }
+      }
+    })
+  )
+
   return function getGroupTangle (groupId, cb) {
+    // return cb(null, { root: "", previous: [] })
```
Comment (mixmix) [inline lib/get-group-tangle.js:37]: Rm

File: lib/get-group-tangle.js [hunk 2]
```diff
@@ -1,18 +1,59 @@
...
     if (!isGroup(groupId)) return cb(new Error(`get-group-tangle expects valid groupId, got: ${groupId}`))
 
+    //  - groupId is a cloaked reference to a group:
+    //     - %asdasdasdasdsasdasda.cloaked
+    //     - info.root is the id of the root message of a group
+    //  - previous is the "tips" we have just calculated and are about to extend on
+    //
```
Comment (mixmix) [inline lib/get-group-tangle.js:44]: Tidy up / rm unneeded notes

File: lib/get-group-tangle.js [hunk 3]
```diff
@@ -29,63 +70,36 @@ module.exports = function GetGroupTangle (server, keystore) {
...
+        // each node should be pruned down to e.g. { key: '%D', previous: ['%B', '%C'] }
 
-    pull(
-      server.backlinks.read({ query }),
-      pull.collect((err, msgs) => {
-        if (err) return cb(err)
+        const reduce = new Reduce(strategy, { nodes })
+        const tipObject = reduce.state
+        // This empty strategy gives reduce.state like { B: {} }
+        // console.log('Reduce state', reduce.state)
```
Comment (mixmix) [inline lib/get-group-tangle.js:96]: Tidy

File: listen.js [hunk 4]
```diff
@@ -36,6 +36,11 @@ function addMember (ssb, emit) {
     ssb.messagesByType({ type: 'group/add-member', private: true, live: true }),
     pull.filter(m => m.sync !== true),
     pull.filter(m => m.value.author !== ssb.id),
+    // pull.through(m => {
+    //   console.log(m.value.content)
+    //   console.log(isValid(m))
+    //   console.log(isValid.error)
+    // }),
```
Comment (mixmix) [inline listen.js:43]: Rm

File: spec/application/application.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+  props: {
+  },
+  addMember: {
+    type: 'object',
+    patternProperties: {
+      [msgIdRegex]: { type: 'integer' }
+    }
+  }
+}
```
Comment (mixmix) [inline spec/application/application.js:32]: This file shouldn't be in this pr.
Please delete and commit

File: test/replicate.test.js [hunk 6]
```diff
@@ -4,7 +4,6 @@ const Keys = require('ssb-keys')
 const { Server, replicate } = require('./helpers')
 
 test('replicate group members', async t => {
-
```
Comment (mixmix) [inline test/replicate.test.js:7]: This whole file might? Be able to be deleted. Need to check

File: test/replicate.test.js [hunk 7]
```diff
@@ -4,7 +4,6 @@ const Keys = require('ssb-keys')
 const { Server, replicate } = require('./helpers')
```
Comment (mixmix) [inline test/replicate.test.js:5]: Really stoked to have this in.
There's some comment and file tidying to be done but other than those I think this is good to go

**Review by mixmix — CHANGES_REQUESTED** [review]
Few things to tidy  but otherwise ready to go

File: package.json [hunk 8]
```diff
@@ -29,15 +29,16 @@
...
-    "ssb-ref": "^2.14.0",
-    "ssb-schema-definitions": "^3.1.0",
-    "ssb-tangle": "^2.1.0"
+    "ssb-keys": "^8.0.1",
+    "ssb-ref": "^2.14.3",
+    "ssb-schema-definitions": "^3.1.0"
   },
   "devDependencies": {
+    "@tangle/reduce": "^2.0.0",
+    "@tangle/strategy": "^1.4.1",
```
[the remark this answers is not in the document]
Comment (CPilbrow) [reply package.json:38]: I don't see anything in the latest version column which is odd. Updated to latest versions and checked tests still pass.

File: lib/get-group-tangle.js [hunk 9]
```diff
@@ -1,18 +1,59 @@
-const getHeads = require('ssb-tangle/graph-tools/get-heads')
 const { isCloakedMsg: isGroup } = require('ssb-ref')
 const pull = require('pull-stream')
+const Reduce = require('@tangle/reduce')
+const Strategy = require('@tangle/strategy')
 
 // for figuring out what "previous" should be for the group
 
 const TANGLE = 'group'
+const strategy = new Strategy({}) //, getBacklinks }), default is node => node.previous
```
[the remark this answers is not in the document]
Comment (CPilbrow) [reply lib/get-group-tangle.js:9]: Sure was

File: listen.js (hunk 4 above)
Comment (CPilbrow) [reply listen.js:43]: Rmed

File: spec/application/application.js (hunk 5 above)
Comment (CPilbrow) [reply spec/application/application.js:32]: Safely deleted. I'm unsure what the file was.

File: test/replicate.test.js (hunk 6 above)
Comment (CPilbrow) [reply test/replicate.test.js:7]: I'm not really sure what's being tested (installReplicate and persistance when restarting server?) so I'll leave it to you to decide.  I do note that I somehow  missed adding live:true to this replicate when I ctrl-shift-F-ed and yet it still seems to pass fine.

Comment (mixmix) [reply test/replicate.test.js:7]: yeah got it. Figured out what this test was for and will leave clearer notes for future explorerz

**Pull request merged**